arangod/Aql/CollectNode.cpp

```diff
@@ -126,8 +128,42 @@ std::unique_ptr<ExecutionBlock> CollectNode::createBlock(
       return std::make_unique<HashedCollectBlock>(&engine, this);
     case CollectOptions::CollectMethod::SORTED:
       return std::make_unique<SortedCollectBlock>(&engine, this);
-    case CollectOptions::CollectMethod::DISTINCT:
-      return std::make_unique<DistinctCollectBlock>(&engine, this);
+    case CollectOptions::CollectMethod::DISTINCT: {
+      ExecutionNode const* previousNode = getFirstDependency();
+      TRI_ASSERT(previousNode != nullptr);
+
+      std::unordered_set<RegisterId> readableInputRegisters;
+      std::unordered_set<RegisterId> writeableInputRegisters;
+
+      std::vector<std::pair<RegisterId, RegisterId>> groupRegisters;
+      for (auto const& p : _groupVariables) {
+        // We know that planRegisters() has been run, so
+        // getPlanNode()->_registerPlan is set up
+        auto itOut = getRegisterPlan()->varInfo.find(p.first->id);
+        TRI_ASSERT(itOut != getRegisterPlan()->varInfo.end());
+
+        auto itIn = getRegisterPlan()->varInfo.find(p.second->id);
+        TRI_ASSERT(itIn != getRegisterPlan()->varInfo.end());
+        TRI_ASSERT((*itIn).second.registerId < ExecutionNode::MaxRegisterId);
+        TRI_ASSERT((*itOut).second.registerId < ExecutionNode::MaxRegisterId);
+        groupRegisters.emplace_back(std::make_pair((*itOut).second.registerId,
+                                                   (*itIn).second.registerId));
+        writeableInputRegisters.insert((*itOut).second.registerId);
+        readableInputRegisters.insert((*itIn).second.registerId);
+      }
+
+      // fill writeable and readable input/output registers TODO (1st & 2nd parameters of executor infos) !!!!!!
+
+      transaction::Methods* trxPtr = _plan->getAst()->query()->trx();
+      DistinctCollectExecutorInfos infos(getRegisterPlan()->nrRegs[previousNode->getDepth()],
+                                         getRegisterPlan()->nrRegs[getDepth()],
+                                         getRegsToClear(), std::move(readableInputRegisters),
+                                         std::move(writeableInputRegisters),
+                                         std::move(groupRegisters), trxPtr);
+
+      return std::make_unique<ExecutionBlockImpl<DistinctCollectExecutor>>(&engine, this,
+                                                                           std::move(infos));
+    }
```

Comment:
As we replace the other collect blocks as well, this should probably be unified as possible, extracted into separate functions, or both. Don't do it now, we will have a better idea how to do it when more blocks are done. Just wanted to mention it early.